Update cmdlets to support operations at MG scope

mentat9 opened the pull request:
- Add -ManagementGroupName and -SubscriptionId parameters to *-AzureRmPolicyDefinition and *-AzureRmPolicySetDefinition
 - Fix Get-AzurePolicyAssignment to support listing at MG scope
 - Fix Get-AzurePolicyAssignment to support retrieving individual assignments at MG scope
 - Add -Effective and -All filter switches to Get-AzurePolicyAssignment to control $filter parameter
 - Remove duplicate function from Assert.ps1
 - Minor refactoring to consolidate duplicated methods into base class
 - Share parameter set names across different cmdlets
 - Fix existing unit tests, add cleanup
 - Add new unit tests for -ManagementGroup and -Subscription parameters
 - Add new unit tests covering all parameter sets of all 12 cmdlets
 - Refactor all parameter help messages into a single file/class
 - Many corrections and improvements to parameter help messages

<!-- DO NOT DELETE THIS TEMPLATE -->

## Description

Update Policy cmdlets for api-version 2018-03-01 and management groups GA

## Checklist

- [X] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [X] The title of the PR is clear and informative
- [X] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [X] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [X] If applicable, the changes made in the PR have proper test coverage
- [X] For public API changes to cmdlets:
    - [X] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [X] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/GetAzurePolicySetDefinition.cs:
```diff
@@ -18,44 +18,71 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
+        [Parameter(ParameterSetName = PolicyCmdletBase.AllFilterParameterSet, Mandatory = false, HelpMessage = PolicyHelpStrings.GetPolicySetDefinitionSubscriptionIdHelp)]
         [ValidateNotNullOrEmpty]
-        public string Name { get; set; }
+        public Guid? SubscriptionId { get; set; }
```
same subId comment as the PolicyDefinition cmdlet

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/NewAzurePolicyAssignment.cs:
```diff
@@ -190,7 +171,7 @@ private JToken GetResource()
             {
                 Name = this.Name,
-                Sku = this.Sku == null? new PolicySku { Name = "A0", Tier = "Free" } : this.Sku.ToDictionary(addValueLayer: false).ToJson().FromJson<PolicySku>(),
+                //Sku = this.Sku == null? new PolicySku { Name = "A0", Tier = "Free" } : this.Sku.ToDictionary(addValueLayer: false).ToJson().FromJson<PolicySku>(),
```
Can remove this line altogether

mentat9 replied:
I'll handle Sku the same as in Set-AzureRmPolicyAssignment. Pending review with the Powershell team, we will either add the [Obsolete] attribute, which requires removing all references to Sku, or we will leave it supported, but change the logic to only send Sku to the backend if it was actually provided as input.

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/NewAzurePolicyDefinition.cs:
```diff
@@ -12,75 +12,85 @@
+        /// </summary>
+        [Parameter(ParameterSetName = PolicyCmdletBase.SubscriptionIdParameterSet, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = PolicyHelpStrings.NewPolicyDefinitionSubscriptionIdHelp)]
+        [ValidateNotNullOrEmpty]
+        public Guid? SubscriptionId { get; set; }
```
same subId comment

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/NewAzurePolicySetDefinition.cs:
```diff
@@ -17,60 +17,73 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
+        /// </summary>
+        [Parameter(ParameterSetName = PolicyCmdletBase.SubscriptionIdParameterSet, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = PolicyHelpStrings.NewPolicySetDefinitionSubscriptionIdHelp)]
+        [ValidateNotNullOrEmpty]
+        public Guid? SubscriptionId { get; set; }
```
same subId comment

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/PolicyCmdletBase.cs:
```diff
@@ -49,6 +74,23 @@ protected PSObject[] GetOutputObjects(string resourceType, params JToken[] resou
+                    var psobject = resource.ToResource().ToPsObject();
+                    psobject.Properties.Add(new PSNoteProperty(resourceType, psobject.Properties["ResourceId"].Value));
+                    return psobject;
+                }).Where(result => (filter == ListFilter.None || string.Equals(((PSObject)result.Properties["Properties"].Value).Properties["PolicyType"].Value.ToString(), filter.ToString(), StringComparison.OrdinalIgnoreCase))).ToArray();
```
Be null safe on PolicyType here in case Azure Stack is running an old version that doesn't have policyType (don't think that is possible but who knows).

mentat9 replied:
Done

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/PolicyCmdletBase.cs:
```diff
@@ -106,5 +148,97 @@ protected JToken GetObjectFromParameter(string parameter)
+        /// </summary>
+        protected string MakePolicyAssignmentId(string scope, string resourceName)
+        {
+            return ResourceIdUtility.GetResourceId(
```
I think there was a bug with this method where it didn't produce the right thing if the assignment was an extension of another resource (i.e. an MG assignment or assignment on an individual resource)

mentat9 replied:
You might be thinking about the ResourceIdUtility.GetResourceName(resourceId) method. Our use of that one definitely got broken by MG scopes, but I didn't modify the method itself because it's used all over the place and I didn't want to break other stuff or spread this PR into other cmdlets. I just changed our cmdlet logic to only use it in a safe way. I didn't see any problems with GetResourceId( ... ), and I verified that one at all scope levels. It's pretty trivial: it just glues stuff containing Name to the end of Scope (at least the way we use it).

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/PolicyCmdletBase.cs:
```diff
@@ -106,5 +148,97 @@ protected JToken GetObjectFromParameter(string parameter)
+        /// <summary>
+        /// Returns the policy or policy set definition resource Id given management group name and/or subscription id and/or policy definition name
+        /// </summary>
+        private string MakeDefinitionId(string managementGroupName, Guid? subscriptionId, string policyDefinitionName, string providerName)
```
"providerName" is the fully qualified resource type (i.e. Microsoft.Authorization/policyDefinitions), right?  Lets change it to fullyQualifedResourceType or something similar.  providerName would mean just "Microsoft.Authorization"

mentat9 replied:
Done

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/RemoveAzurePolicyDefinition.cs:
```diff
@@ -15,47 +15,53 @@
+        /// </summary>
+        [Parameter(ParameterSetName = PolicyCmdletBase.SubscriptionIdParameterSet, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = PolicyHelpStrings.RemovePolicyDefinitionSubscriptionIdHelp)]
+        [ValidateNotNullOrEmpty]
+        public Guid? SubscriptionId { get; set; }
```
same subId comment as above

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/RemoveAzurePolicySetDefinition.cs:
```diff
@@ -15,47 +15,53 @@
+        /// </summary>
+        [Parameter(ParameterSetName = PolicyCmdletBase.SubscriptionIdParameterSet, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = PolicyHelpStrings.RemovePolicySetDefinitionSubscriptionIdHelp)]
+        [ValidateNotNullOrEmpty]
+        public Guid? SubscriptionId { get; set; }
```
same subId comment as above

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/SetAzurePolicyAssignment.cs:
```diff
@@ -17,77 +17,68 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
         /// Gets or sets the policy assignment not scopes parameter.
         /// </summary>
-        [Parameter(ParameterSetName = SetAzurePolicyAssignmentCmdlet.PolicyAssignmentNameParameterSet, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The policy assignment not scopes.")]
+        [Parameter(ParameterSetName = PolicyCmdletBase.NameParameterSet, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = PolicyHelpStrings.SetPolicyAssignmentNotScopesHelp)]
```
This seems like an oversight that NotScope wasn't allowed in IdParameterSet.  I don't think this would qualify as a breaking change

mentat9 replied:
Nice catch, I agree.

mentat9 replied:
Done

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/SetAzurePolicyAssignment.cs:
```diff
@@ -131,9 +122,9 @@ private JToken GetResource(string resourceId, string apiVersion)
-                Sku = this.Sku != null
-                    ? this.Sku.ToDictionary(addValueLayer: false).ToJson().FromJson<PolicySku>()
-                    : (resource.Sku == null ? new PolicySku { Name = "A0", Tier = "Free" } : resource.Sku.ToJson().FromJson<PolicySku>()),
+                //Sku = this.Sku != null
```
Since SKU still exists in the API for now I think this should persist it if the user is explicitly providing it in the cmdlet.  It should not include it if they did not include it

mentat9 replied:
Sure, I can put it back. The reason I removed it is because I wanted to mark the parameter [Obsolete]. But doing so means any reference to the parameter causes a build break. So we have to choose between obsoleting it and keeping it. I guess you still want to keep it? Also, I guess you are still OK with throwing it away for New-AzureRmPolicyAssignment?

pilor replied:
ah, if you are moving towards making it obsolete here then its fine.  Not sure how long that process takes (i.e. how many PS releases the obsolete tag is around until the parameter can be removed)

mentat9 replied:
OK, I'll add it back to the PR and we can cover that in the design meeting as well. I don't know what the obsoleting process looks like either.

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/SetAzurePolicyAssignment.cs:
```diff
@@ -131,9 +122,9 @@ private JToken GetResource(string resourceId, string apiVersion)
+                //Sku = this.Sku != null
+                //    ? this.Sku.ToDictionary(addValueLayer: false).ToJson().FromJson<PolicySku>()
+                //    : (resource.Sku == null ? new PolicySku { Name = "A0", Tier = "Free" } : resource.Sku.ToJson().FromJson<PolicySku>()),
                 Properties = new PolicyAssignmentProperties
```
Does this blow away parameters and metadata?  Seems broken

mentat9 replied:
I verified it is indeed broken in the current cmdlet. It's also the RC of a reported issue. I'll fix it in my next update.

mentat9 replied:
Done

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/PolicyHelpStrings.cs:
```diff
@@ -0,0 +1,99 @@
+        public const string SetPolicySetDefinitionIdHelp = "The fully qualified policy set definition ID to get, including the subscription or management group. e.g. /providers/Microsoft.Management/managementGroups/{managementGroup}/providers/Microsoft.Authorization/policySetDefinitions/{policySetDefinitionName}.";
+        public const string SetPolicySetDefinitionDisplayNameHelp = "The display name of the updated policy set definition.";
+        public const string SetPolicySetDefinitionDescriptionHelp = "The description of the updated policy set definition.";
+        public const string SetPolicySetDefinitionPolicyDefinitionHelp = "The policy definition of the updated policy set definition. This can either be a path to a file name containing the policy definitions, or the policy definition as a string.";
```
use "definitions" (plural) in the help string

mentat9 replied:
Done. Similarly changed parameter => parameters

pilor commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/SetAzurePolicySetDefinition.cs:
```diff
@@ -20,64 +20,69 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
+        public Guid? SubscriptionId { get; set; }
+
         /// <summary>
         /// Executes the cmdlet.
```
Should support updating metadata like the policyDefinition cmdlet does

pilor replied:
Also parameters

mentat9 replied:
Done

mentat9 commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/GetAzurePolicyAssignment.cs:
```diff
@@ -18,60 +18,60 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
         /// </summary>
-        [Parameter(ParameterSetName = GetAzurePolicyAssignmentCmdlet.PolicyAssignmentIdParameterSet, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The fully qualified policy assignment Id, including the subscription. e.g. /subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}")]
-        [Parameter(ParameterSetName = GetAzurePolicyAssignmentCmdlet.PolicyAssignmentNameParameterSet, Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "The fully qualified policy assignment Id, including the subscription. e.g. /subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}")]
+        [Parameter(ParameterSetName = PolicyCmdletBase.PolicyDefinitionIdFilterParameterSet, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = PolicyHelpStrings.GetPolicyDefinitionFilterHelp)]
```
Hmm, what does it mean to provide and ID and a filter in the same command?

pilor replied:
its not very helpful, I agree.  It technically lets you check if the policyAssignment is for the policyDefinitionId.  Either way, this should still be usable in both the Name and Id parameter sets so we don't break any scripts

mentat9 replied:
AFAICT, those combinations were allowed but meaningless. The existing cmdlet silently ignores the -PolicyDefinitionId parameter if -Id or -Name is provided. Although making that into a parameter set error is a breaking change, I think we should make it anyway. Anyone using that combination today isn't getting back what they think they are getting back.

pilor replied:
sounds good

mentat9 commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/GetAzurePolicyDefinition.cs:
```diff
@@ -18,44 +18,71 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
+        [Parameter(ParameterSetName = PolicyCmdletBase.AllFilterParameterSet, Mandatory = false, HelpMessage = PolicyHelpStrings.GetPolicyDefinitionSubscriptionIdHelp)]
         [ValidateNotNullOrEmpty]
-        public string Name { get; set; }
+        public Guid? SubscriptionId { get; set; }
```
Couldn't this cmdlet also used to list/get policies from other subscriptions? That's the scenario for this parameter anyway. If it's not a thing, I'll remove the parameter.

pilor replied:
The user should change their context with Select-AzureSubscription in order to list from a different subscription

mentat9 replied:
Leaving -SubscriptionId in place for all cmdlets pending design review discussion

mentat9 commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/GetAzurePolicyDefinition.cs:
```diff
@@ -18,44 +18,71 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
+        /// Gets or sets the all switch.
+        /// </summary>
+        [Parameter(ParameterSetName = PolicyCmdletBase.AllFilterParameterSet, Mandatory = true, HelpMessage = PolicyHelpStrings.GetPolicyDefinitionAllFilterHelp)]
+        public SwitchParameter All { get; set; }
```
Isn't that a normal PS thing? I thought it's a good practice that if you have switches that override a default, you should also have a switch to explicitly specify the default. Agree it's not required for full functionality.

pilor replied:
I don't know, just seems weird.  If its normal PS then its fine

mentat9 replied:
I've requested a design meeting (and included you) to go over this PR with someone from the PS team. We should be able to get clarity on this point. Design meeting request status is here: [https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/74](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/74)

mentat9 commented on src/ResourceManager/Resources/ChangeLog.md:
```diff
@@ -18,6 +18,31 @@
+* Update Get/New/Remove/Set-AzureRmPolicySetDefinition cmdlets
+    - Add -ManagementGroupName parameter to apply operations to a given management group
+    - Add -SubscriptionId parameter to apply operations to a given subscription
+* Minor refactoring to remove duplicated code from Policy cmdlets
```
Done

mentat9 commented:
I don't know how to fix this [built break](https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/5123/). "Modified files were found with no update to their change log. Please add a snippet to the affected modules' change log.". There are many hundreds of files listed, including the ones I changed, but AFAICT, all of the ones I changed should be covered by my update to changelog.md. The one exception is BreakingChangeIssues.csv, which doesn't seem to have a changelog.md that covers it. Can someone advise how to fix this? Even if I do something to eliminate the failure for my files, what about all of the others?

vladimir-shcherbakov commented:
@mentat9 
Is it possible for you to do a rebase for your commits? The goal here is to get the latest preview branch version and add your commits on top of it.

mentat9 commented:
OK, just completed rebase off of main/preview. Maybe this will clean up the build...

vladimir-shcherbakov commented:
Looks good!

mentat9 commented:
Build break looks about the same to me. Any ideas?

vladimir-shcherbakov commented:
@mentat9 
The idea is that after a rebase there should only your commits left in the PR. All other commits should go away. Now there are 155 commits in the PR and most of them are not yours.

cormacpayne commented:
@mentat9 Hey Chris, our guide for rebasing can be found [here](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/cleaning-up-commits.md#rebasing).

You'll run the following commands:

(1) `git fetch upstream`

Replace `upstream` with whatever name you have given the `Azure/azure-powershell` remote on your machine

(2) `git pull upstream preview`

Pull the latest content from the `preview` branch into your local working branch

(3) `git reset --soft upstream/preview`

"Uncommit" all of the content on your local machine that's different from what's on the `preview` branch

(4) <Commit your changes>

Commit your local changes

(5) `git push origin preview -f`

**Force** push your changes to your remote -- this should leave you with a single commit (unless you did multiple commits in step (4) )

vladimir-shcherbakov commented:
Try this:
```git
git remote add upstream https://github.com/Azure/azure-powershell.git
git fetch upstream 
git merge upstream /preview
git reset --soft upstream /preview
```

mentat9 commented:
Please be sure and examine the design review as well: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/74

mentat9 commented on .gitignore:
```diff
@@ -26,6 +26,7 @@ app.config
 *.userprefs
 *.sln.docstates
 .vscode
+*.userosscache
```
It's a per-user file created by Visual Studio. These should never be checked in.

mentat9 commented on src/ResourceManager/Common/Commands.ScenarioTests.ResourceManager.Common/Assert.ps1:
```diff
@@ -191,31 +191,6 @@ function Assert-False
-#    param [ScriptBlock] $script : The script to execute
-#    param [string] $message     : The message to return if the given script does not return false
-####################
-function Assert-False
```
I originally had a comment explaining that change. It must have gotten squashed out later. The Assert-False function was in the file twice identically. I hate leaving stuff like that, but if I have to put it back to get the PR approved, I will do so in a heartbeat.

cormacpayne replied:
Got it, didn't know that it was a duplicated function. Ignore this comment.

mentat9 commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/GetAzurePolicyAssignment.cs:
```diff
@@ -18,60 +18,60 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
     /// Gets the policy assignment.
     /// </summary>
-    [Cmdlet(VerbsCommon.Get, "AzureRmPolicyAssignment", DefaultParameterSetName = GetAzurePolicyAssignmentCmdlet.ParameterlessSet), OutputType(typeof(PSObject))]
+    [Cmdlet(VerbsCommon.Get, "AzureRmPolicyAssignment", DefaultParameterSetName = PolicyCmdletBase.DefaultParameterSet), OutputType(typeof(PsPolicyAssignment))]
```
Did you read the design review comments here: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/74? This change was rather insistently requested by @vladimir-shcherbakov, despite my initial objections. One concern I mentioned was that it could be a breaking change, although @vladimir-shcherbakov was confident it was fine. To be sure, I prototyped the change and did some testing. Upon study it appears that the only thing it is likely to break was a script using strict semantics and testing property existence to determine "present or not", rather than a check for $null. In non-strict mode both results are the same. Of course, since the underlying type is different, it's always possible to concoct an example it would break (e.g. explicit check for PSObject type), but those cases don't seem realistic to me. Am I missing other important breaking scenarios?

mentat9 replied:
>I might have missed it, but where are we returning a PsPolicyAssignment object rather than a PSObject now?

If your question is asking specifically where in the code the new object types are getting constructed, take a look here: https://github.com/mentat9/azure-powershell/blob/preview/src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/PolicyCmdletBase.cs in method GetOutputPolicyAssignments()

cormacpayne replied:
Sorry, I should've been more explicit about the kind of breaking changes we look for in this case -- this change from `PSObject` to `PsPolicyAssignment` might not introduce a structural breaking change (one that would cause an existing script to throw) but could introduce a behavioral breaking change (script continues to work, but users don't get back the same result from one version to the next). My concern is that users have scripts that use properties from the object returned from `Get-AzureRmPolicyAssignment` that may not exist in the `PsPolicyAssignment` object, but we may need to do some more investigation to determine what values are returned from the server and if any of them don't show up as a property for `PsPolicyAssignment` (and for other types converted from `PSObject` in this PR). We can talk offline about this some more if needed since this is a change we would want to avoid making for the stable `AzureRM.Resources` module if it means sacrificing the piping scenario.

mentat9 replied:
The wrapper classes I created specifically include every property that comes back from the server. Let me know how to follow this up with your team.

mentat9 commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/GetAzurePolicyAssignment.cs:
```diff
@@ -18,60 +18,60 @@ namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation
+        /// Gets or sets the all switch
+        /// </summary>
+        [Parameter(ParameterSetName = PolicyCmdletBase.AllFilterParameterSet, Mandatory = true, ValueFromPipelineByPropertyName = true, HelpMessage = PolicyHelpStrings.GetPolicyAssignmentAllFilterHelp)]
+        public SwitchParameter All { get; set; }
```
I think it would be confusing, since the default without a filter is -Effective which gives less than -All. Basically, that approach would give you more items when you apply a filter than when you don't. The default can't be changed without a big breaking change.

cormacpayne replied:
Just to make sure I understand, the current behavior without either of these switch parameter is "filter by effective", and the cmdlet doesn't return all policy assignments? If so, is there a difference between `Get-AzureRmPolicyAssignment` and `Get-AzureRmPolicyAssignment -Effective`?

mentat9 replied:
That's correct, so there's no difference. I included -Effective so there is a name for what happens by default.

cormacpayne replied:
I think then rather exposing this parameter that has no effect, we rename the `-All` parameter to describe what type of policy assignments are being returned (if they have a name) in addition to the effective ones returned by default. The parameter would then be renamed to be `-IncludeFoo`, where Foo is the name of these additional assignments.

mentat9 replied:
OK. Unfortunately there isn't really a name for that set. I'll remove -Effective and try to think of a better name for -All. So far, the best I have is -IncludeDescendents.

mentat9 commented on src/ResourceManager/Resources/Commands.Resources/Commands.Resources.csproj:
```diff
@@ -193,6 +193,87 @@
       <Link>AzureRM.Resources.psd1</Link>
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Include="help\Add-AzureRmADGroupMember.md" />
```
OK

mentat9 commented on src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmPolicyAssignment.md:
```diff
@@ -1,4 +1,4 @@
-﻿---
```
Sure

mentat9 commented on src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmPolicyAssignment.md:
````diff
@@ -13,22 +13,43 @@ Gets policy assignments.
-### GetPolicyAssignmentName
+### NameParameterSet
 ```
-Get-AzureRmPolicyAssignment [-Name <String>] -Scope <String> [-PolicyDefinitionId <String>]
````
It's not worth the work of obsoleting it. I'll just bring it back in, and it will remain a no-op.

mentat9 replied:
It turns out that I can't bring that parameter back in without removing some new functionality I added. This update adds the ability to give -Name without -Scope (previously that was not allowed). In order to return -PolicyDefinitionId to the way it was, I need to remove that functionality. So to avoid a breaking change to a combination of parameters that doesn't make sense (and doesn't even work), I have to actually put back a bug that was present in preview cmdlets and remove new functionality from the GA cmdlets. I don't see that as a win for the customer. Given the situation, would you reconsider this point?

cormacpayne replied:
What was the previous behavior of the cmdlet if the user provided the `-PolicyDefinitionId` parameter? If it's the case that the cmdlet would throw an exception everytime, then it is fine to remove this parameter since no user would have an existing script with this parameter that worked properly. If it didn't throw an exception (even if its value wasn't being used), then it's a breaking change, which is something that we cannot allow until our next breaking change release. I agree that exposing new functionality to customers is a much better experience than keeping functionality that doesn't work as intended, but our goal is to avoid making breaking changes.

Looking at the changes and additions you are making to the parameter sets, it seems that you can bring back the `-PolicyDefinitionId` parameter to the `NameParameterSet` (to avoid the breaking change) and remove the `-Scope` parameter from the `PolicyDefinitionIdFilterParameterSet` set to keep the new functionality (the `-PolicyDefinitionId` + `-Scope` scenario should now be handled in the `NameParameterSet` set). This should allow you to handle all previous cases as well as the new cases you are introducing. Do you see any issue with this?

mentat9 replied:
Yeah, that's what I thought too, but that combination doesn't allow -PolicyDefinitionId with no other parameters, which needs to be valid input.

mentat9 replied:
The previous behavior was that the value for -PolicyDefinitionId was silently ignored, so the results back were the same whether you provided -PolicyDefinitionId or not, and no matter what nonsense value you might have provided.

cormacpayne replied:
Are we able to remove the `PolicyDefinitionIdFilterParameterSet` set all together and bring back the `-PolicyDefinitionId` parameter to the `NameParameterSet` as an optional parameter? That allows you to provide `-PolicyDefinitionId` with no other parameters, and if either of `-Scope` or `-Name` are provided, you can ignore the value of `-PolicyDefinitionId` like you did previously

mentat9 replied:
No. PolicyDefinitionId has a parameter set because it needs to be mutually exclusive from -All and -Effective. Even after removing -Effective, it still needs to be exclusive from -All (or -IncludeDescendents if it gets renamed). I could enforce that by explicit check instead of PS parameter sets, but I'm guessing you won't like that approach.

mentat9 commented on src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmPolicySetDefinition.md:
````diff
@@ -12,42 +12,74 @@ Gets policy set definitions.
-### GetBySubscription (Default)
+### NameParameterSet (Default)
+```
+Get-AzureRmPolicySetDefinition [-Name <String>] [-ApiVersion <String>] [-Pre]
````
Yes.

cormacpayne replied:
👍

mentat9 commented on src/ResourceManager/Resources/Commands.Resources/help/Set-AzureRmPolicyDefinition.md:
````diff
@@ -13,76 +13,98 @@ Modifies a policy definition.
+### NameParameterSet (Default)
 ```
-Set-AzureRmPolicyDefinition [-DisplayName <String>] [-Description <String>] [-Policy <String>]
+Set-AzureRmPolicyDefinition -Name <String> [-DisplayName <String>] [-Description <String>] [-Policy <String>]
````
The existing cmdlet crashes if you don't give -Name, so we aren't breaking anyone.

cormacpayne replied:
Awesome, thanks for clarification. Ignore this comment.

mentat9 commented on src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmPolicyDefinition.md:
````diff
@@ -144,31 +165,130 @@ Specifies the name of the policy definition that this cmdlet gets.
+Accept wildcard characters: False
+```
+
+### -SubscriptionId
````
That example doesn't work for me:

```
[D:\git\repos\azure-powershell\src\ResourceManager\Resources]-> Get-AzureRmVM -DefaultProfile (Get-AzureRmContext -Name 'AIMES Deployment Test')
Get-AzureRmContext : Cannot validate argument on parameter 'Name'. The argument "AIMES Deployment Test" does not belong to the set "Azure Governance INT -
d0610b27-9663-4c05-89f8-5b4be01e86a5" specified by the ValidateSet attribute. Supply an argument that is in the set and then try the command again.
At line:1 char:57
+ ... VM -DefaultProfile (Get-AzureRmContext -Name 'AIMES Deployment Test')
+                                                  ~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Get-AzureRmContext], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationError,Microsoft.Azure.Commands.Profile.GetAzureRMContextCommand
```
Despite whether that works or not, this is qualitatively different than the -DefaultProfile parameter. When this parameter (or -ManagementGroupName) is used, policies aren't resources "in a" user's profile/subscription, they are instead "applied to" a subscription or management group from outside. It doesn't make logical sense that a user would have to enter the context of each subscription they want to apply a policy to. All they need is the proper RBAC permissions on the subscription they are using. This update is to bring the cmdlets into parity with the portal UI, which supports applying policies to management groups and subscriptions without switching/entering contexts. You can see that UI in action here: https://ms.portal.azure.com/#blade/Microsoft_Azure_Policy/PolicyMenuBlade/Assignments by clicking on the [...] button next to Scope.

mentat9 commented on src/ResourceManager/Resources/Commands.Resources/help/Get-AzureRmPolicyAssignment.md:
````diff
@@ -13,22 +13,43 @@ Gets policy assignments.
-### GetPolicyAssignmentId
+### IdParameterSet
 ```
-Get-AzureRmPolicyAssignment -Id <String> [-PolicyDefinitionId <String>] [-ApiVersion <String>] [-Pre]
````
Same

mentat9 commented:
@cormacpayne, I've responded to all of your feedback and applied updates. Please take another look as soon as you can.

mentat9 commented:
Per our meeting Friday, I pushed updates for all remaining feedback on the "new/preview/future" version of the cmdlets. This PR is ready to merge to that branch (once validation passes). I will create a separate PR for the "stable" version of the cmdlets when it is ready (should be later today).

cormacpayne commented:
@mentat9 Can you please fix the merge conflict? We have moved ChangeLog/psd1 files from `src/ResourceManager/` to `src/ResourceManager/{{service}}/Commands.{{service}}`, so you will need to move your ChangeLog/psd1 updates there.

mentat9 commented:
Merge conflict resolved.

cormacpayne commented:
@mentat9 Hey Chris, I merged in your non-breaking PR into the `preview` branch, and have created an `AzureRM.Resources-preview` branch that this PR is now targeting. Would you mind pulling in the latest from this branch and resolving the merge conflicts? I assume most conflicts will be easily resolved since they are the same changes 😀

mentat9 commented:
Conflicts resolved.

cormacpayne commented:
PR to merge the latest from the `preview` branch (post release) into `AzureRM.Resources-preview`: https://github.com/Azure/azure-powershell/pull/6564

Once this is merged, we can get this PR into that branch, do the necessary versioning, and start the release process

cormacpayne commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/PSPolicyDefinition.cs:
```diff
@@ -0,0 +1,33 @@
+﻿namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.Policy
```
@mentat9 please add a license header to this file

cormacpayne commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/PsPolicyAssignment.cs:
```diff
@@ -0,0 +1,37 @@
+﻿namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.Policy
```
@mentat9 same comment about license header

cormacpayne commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/PsPolicySetDefinition.cs:
```diff
@@ -0,0 +1,33 @@
+﻿namespace Microsoft.Azure.Commands.ResourceManager.Cmdlets.Implementation.Policy
```
@mentat9 same comment about license header

cormacpayne commented on src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Policy/SetAzurePolicyDefinition.cs:
```diff
@@ -94,6 +94,12 @@ public class SetAzurePolicyDefinitionCmdlet : PolicyCmdletBase
+        /// <summary>
+        /// Gets or sets the policy definition input object parameter.
+        /// </summary>
+        [Parameter(ParameterSetName = PolicyCmdletBase.InputObjectParameterSet, Mandatory = true, ValueFromPipeline = true, ValueFromPipelineByPropertyName = true, HelpMessage = PolicyHelpStrings.SetPolicyDefinitionInputObjectHelp)]
```
@mentat9 for the `-InputObject` parameters you are adding, you can remove the `ValueFromPipelineByPropertyName` attribute (you only need `ValueFromPipeline`)

mentat9 commented:
Updated to resolve all current feedback.
